polkadot/runtime/common/src/paras_registrar/mod.rs

```diff
@@ -642,6 +794,79 @@ impl<T: Config> Pallet<T> {
 		Ok(())
 	}
 
+	/// Schedules a code upgrade for a parachain.
+	///
+	/// If `free_upgrade` is set to true, there won't be any extra deposit or fees charged
+	/// for scheduling the code upgrade.
+	///
+	/// If the size of the validation is reduced and the upgrade is successful the caller will be
+	/// eligible for receiving back a portion of their deposit that is no longer required.
+	fn do_schedule_code_upgrade(
+		para: ParaId,
+		new_code: ValidationCode,
+		free_upgrade: bool,
+	) -> DispatchResult {
+		// Before doing anything we ensure that a code upgrade is allowed at the moment for the
+		// specific parachain.
+		ensure!(paras::Pallet::<T>::can_upgrade_validation_code(para), Error::<T>::CannotUpgrade);
+
+		let head =
+			paras::Pallet::<T>::para_head(para).map_or(Err(Error::<T>::NotRegistered), Ok)?;
+
+		let per_byte_fee = T::DataDepositPerByte::get();
+		let new_deposit = T::ParaDeposit::get()
+			.saturating_add(per_byte_fee.saturating_mul((head.0.len() as u32).into()))
+			.saturating_add(per_byte_fee.saturating_mul((new_code.0.len() as u32).into()));
+
+		let mut info = Paras::<T>::get(para).map_or(Err(Error::<T>::NotRegistered), Ok)?;
+
+		let billing_account = info.billing_account.clone();
+		ensure!(billing_account.clone().is_some(), Error::<T>::BillingAccountNotSet);
+		let billing_account =
+			billing_account.expect("Ensured above that the billing account is set to some; qed");
+
+		let current_deposit = info.deposit;
+
+		if !free_upgrade {
+			<T as Config>::Currency::withdraw(
+				&billing_account,
+				T::UpgradeFee::get(),
+				WithdrawReasons::FEE,
+				ExistenceRequirement::KeepAlive,
+			)?;
+
+			let additional_deposit = new_deposit.saturating_sub(current_deposit);
+			<T as Config>::Currency::reserve(&billing_account, additional_deposit)?;
+
+			// Update the deposit to the new appropriate amount.
+			info.deposit = new_deposit;
+		}
+
+		if current_deposit > new_deposit {
+			// The billing account should receive a refund if the current deposit exceeds the new
+			// required deposit, even if they did not initiate the upgrade.
+			//
+			// The excess deposit will be refunded to the caller upon the success of the code
+			// upgrade.
+			//
+			// The reason why the deposit is not instantly refunded is that scheduling a code
+			// upgrade doesn't guarantee the success of it.
+			//
+			// If we returned the deposit here, a possible attack scenario would be to register
+			// the validation code of a parachain and then schedule a code upgrade to set the
+			// code to an empty blob. In such a case, the pre-checking process would fail, so
+			// the old code would remain on-chain even though there is no deposit to cover it.
+
+			info.pending_refund = Some(current_deposit.saturating_sub(new_deposit));
+		}
+
+		Paras::<T>::insert(para, info);
+		runtime_parachains::schedule_code_upgrade::<T>(para, new_code, SetGoAhead::No)?;
```

Comment:
I just realized that none of this works here. The most common way to do a runtime upgrade is by the parachain itself, which calls [this function](https://github.com/paritytech/polkadot-sdk/blob/e1c033ebe15e6de5ee6a5a81eab2dd213f09d574/polkadot/runtime/parachains/src/inclusion/mod.rs#L885) directly. 

Hence even with your PR, parachains could still upgrade for free and also their deposits won't be adjusted. On top of that the check for root origin to determine special rights for system parachains will also not work this way.

Sorry, did not spot this either initially. I just realized this, because I wanted to check how the origin is even set if the upgrade is initiated by the parachain itself .... result: It is not.